app/bundles/LeadBundle/Model/CompanyModel.php

```diff
@@ -0,0 +1,490 @@
+use Mautic\LeadBundle\Entity\Company;
+use Mautic\LeadBundle\Entity\CompanyLead;
+use Mautic\LeadBundle\Entity\CompanyLeadRepository;
+use Mautic\LeadBundle\Event\CompanyBuilderEvent;
+use Mautic\LeadBundle\Event\LeadChangeCompanyEvent;
+use Mautic\LeadBundle\LeadEvents;
+use Symfony\Component\HttpKernel\Exception\MethodNotAllowedHttpException;
+use Symfony\Component\HttpFoundation\Session\Session;
+
+/**
+ * Class CompanyModel
+ */
+class CompanyModel extends CommonFormModel
+{
+    /**
+     * @var Session
+     */
+    protected $session;
+
+    /**
+     * @var FieldModel
+     */
+    protected $leadFieldModel;
+    /**
+     * PointModel constructor.
+     *
+     * @param Session $session
+     *
+     */
+    public function __construct(FieldModel $leadFieldModel, Session $session)
+    {
+        $this->leadFieldModel = $leadFieldModel;
+        $this->session = $session;
+    }
+    /**
+     * {@inheritdoc}
+     *
+     * @return \Mautic\LeadBundle\Entity\CompanyRepository
+     */
+    public function getRepository()
+    {
+        return $this->em->getRepository('MauticLeadBundle:Company');
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return \Mautic\LeadBundle\Entity\CompanyLeadRepository
+     */
+    public function getCompanyLeadRepository()
+    {
+        return $this->em->getRepository('MauticLeadBundle:CompanyLead');
+    }
+
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getPermissionBase()
+    {
+        return 'company:companies';
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return string
+     */
+    public function getNameGetter()
+    {
+        return "getPrimaryIdentifier";
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws MethodNotAllowedHttpException
+     */
+    public function createForm($entity, $formFactory, $action = null, $options = array())
+    {
+        if (!$entity instanceof Company) {
+            throw new MethodNotAllowedHttpException(array('Company'));
+        }
+        if (!empty($action)) {
+            $options['action'] = $action;
+        }
+        return $formFactory->create('company', $entity, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return Company|null
+     */
+    public function getEntity($id = null)
+    {
+        if ($id === null) {
+            return new Company();
+        }
+
+        return parent::getEntity($id);
+    }
+    
+    /**
+     *
+     * @return mixed
+     */
+    public function getUserCompanies()
+    {
+        $user  = (!$this->security->isGranted('lead:leads:viewother')) ?
+            $this->factory->getUser() : false;
+        $companies = $this->em->getRepository('MauticLeadBundle:Company')->getCompanies($user);
+
+        return $companies;
+    }
+
+    /**
+     * Reorganizes a field list to be keyed by field's group then alias
+     *
+     * @param $fields
+     * @return array
+     */
+    public function organizeFieldsByGroup($fields)
```

Comment:
Let's consider using Trait for this other verbatim methods from LeadModel with the minor tweaks needed to customize.